aya/src/programs/mod.rs

```diff
@@ -805,6 +818,61 @@ impl_fd!(
     CgroupDevice,
 );
 
+/// Defines which [`Program`]s support the kernel's
+/// generic multi-prog API.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 6.6.
+pub trait MprogProgram {
+    /// Returns a borrowed reference to the file descriptor of a given
+    /// [`Program`] which has support for the kernel's generic multi-prog API.
+    fn fd(&self) -> Result<BorrowedFd<'_>, ProgramError>;
+}
+
+macro_rules! impl_mprog_fd {
+    ($($struct_name:ident),+ $(,)?) => {
+        $(
+            impl MprogProgram for $struct_name {
+                /// Returns the a borrowed reference file descriptor of this Program.
+                fn fd(&self) -> Result<BorrowedFd<'_>, ProgramError> {
+                    Ok(self.fd()?.as_fd())
+                }
+            }
+        )+
+    }
+}
+
+impl_mprog_fd!(SchedClassifier,);
+
+/// Defines which [`Link`]s support the kernel's
+/// generic multi-prog API.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 6.6.
+pub trait MprogLink {
+    /// Returns a borrowed reference to the file descriptor of a given
+    /// [`Link`] which has support for the kernel's generic multi-prog API.
+    fn fd(&self) -> Result<BorrowedFd<'_>, LinkError>;
+}
+
+macro_rules! impl_mproglink_fd {
+    ($($struct_name:ident),+ $(,)?) => {
+        $(
+            impl MprogLink for $struct_name {
+                /// Returns the a borrowed reference file descriptor of this Program.
+                fn fd(&self) -> Result<BorrowedFd<'_>, LinkError> {
+                    let link: &FdLink = self.try_into()?;
+                    Ok(link.fd.as_fd())
+                }
+            }
+        )+
+    }
+}
+
+impl_mproglink_fd!(SchedClassifierLink);
```

Comment:
What other programs support this?

Reply:
None yet however it is being toted as a the "future" of multi-attach ordering mechanisms https://github.com/torvalds/linux/blob/master/include/linux/bpf_mprog.h#L8